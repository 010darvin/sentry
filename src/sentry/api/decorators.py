from functools import wraps

from rest_framework.request import Request
from rest_framework.response import Response

from sentry.api.exceptions import EmailVerificationRequired, SudoRequired
from sentry.models import ApiKey
from sentry.models.apitoken import is_api_token_auth
from sentry.models.user import User
from sentry.services.hybrid_cloud.user import APIUser


def is_considered_sudo(request):
    # Right now, only password reauthentication (django-sudo) is supported,
    # so if a user doesn't have a password (for example, only has github auth)
    # then we shouldn't prompt them for the password they don't have.
    return (
        request.is_sudo()
        or isinstance(request.auth, ApiKey)
        or is_api_token_auth(request.auth)
        or request.user.is_authenticated
        and not request.user.has_usable_password()
    )


def sudo_required(func):
    @wraps(func)
    def wrapped(self, request: Request, *args, **kwargs) -> Response:
        # If we are already authenticated through an API key we do not
        # care about the sudo flag.
        if not is_considered_sudo(request):
            # TODO(dcramer): support some kind of auth flow to allow this
            # externally
            raise SudoRequired(request.user)
        return func(self, request, *args, **kwargs)

    return wrapped


def email_verification_required(func):
    @wraps(func)
    def wrapped(self, request: Request, *args, **kwargs) -> Response:
        if isinstance(request.user, User):
            if not request.user.get_verified_emails().exists():
                raise EmailVerificationRequired(request.user)
        elif isinstance(request.user, APIUser):
            has_verified_email = False
            for email in request.user.useremails:
                if email.is_verified:
                    has_verified_email = True
                    break
            if not has_verified_email:
                raise EmailVerificationRequired(request.user)
        else:
            raise NotImplementedError(
                "email_verification_required doesn't handle this type of user input"
            )
        return func(self, request, *args, **kwargs)

    return wrapped
