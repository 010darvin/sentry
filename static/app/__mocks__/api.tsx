import isEqual from 'lodash/isEqual';

import type {HandleRequestErrorOptions, RequestOptions, ResponseMeta} from 'sentry/api';
import ModalStore from 'sentry/stores/modalStore';

import {SUDO_REQUIRED, SUPERUSER_REQUIRED} from '../constants/apiErrorCodes';

export class Request {}

export const initApiClientErrorHandling = () => false;
export const hasProjectBeenRenamed = (_args: any[]) => false;

const respond = (isAsync: boolean, fn?: Function, ...args: any[]): void => {
  if (!fn) {
    return;
  }

  if (isAsync) {
    setTimeout(() => fn(...args), 1);
    return;
  }

  fn(...args);
};

type FunctionCallback<Args extends any[] = any[]> = (...args: Args) => void;

/**
 * Callables for matching requests based on arbitrary conditions.
 */
interface MatchCallable {
  (url: string, options: RequestOptions): boolean;
}

type ResponseType = ResponseMeta & {
  body: any;
  callCount: 0;
  headers: Record<string, string>;
  match: MatchCallable[];
  method: string;
  statusCode: number;
  url: string;
};

type MockResponse = [resp: ResponseType, mock: jest.Mock];

/**
 * Compare two records. `want` is all the entries we want to have the same value in `check`
 */
function compareRecord(want: Record<string, any>, check: Record<string, any>): boolean {
  for (const key in want) {
    if (!isEqual(check[key], want[key])) {
      return false;
    }
  }
  return true;
}

afterEach(() => {
  // if any errors are caught we console.error them
  const errors = Object.values(Client.errors);
  if (errors.length > 0) {
    for (const err of errors) {
      // eslint-disable-next-line no-console
      console.error(err);
    }
    Client.errors = {};
  }
});

class Client implements Client {
  static mockResponses: MockResponse[] = [];

  static mockAsync = false;

  static clearMockResponses() {
    Client.mockResponses = [];
  }

  /**
   * Create a query string match callable.
   *
   * Only keys/values defined in `query` are checked.
   */
  static matchQuery(query: Record<string, any>): MatchCallable {
    const queryMatcher: MatchCallable = (_url, options) => {
      return compareRecord(query, options.query ?? {});
    };

    return queryMatcher;
  }

  /**
   * Create a data match callable.
   *
   * Only keys/values defined in `data` are checked.
   */
  static matchData(data: Record<string, any>): MatchCallable {
    const dataMatcher: MatchCallable = (_url, options) => {
      return compareRecord(data, options.data ?? {});
    };

    return dataMatcher;
  }

  // Returns a jest mock that represents Client.request calls
  static addMockResponse(response: Partial<ResponseType>) {
    const mock = jest.fn();

    Client.mockResponses.unshift([
      {
        url: '',
        status: 200,
        statusCode: 200,
        statusText: 'OK',
        responseText: '',
        responseJSON: '',
        body: '',
        method: 'GET',
        callCount: 0,
        match: [],
        ...response,
        headers: response.headers ?? {},
        getResponseHeader: (key: string) => response.headers?.[key] ?? null,
        rawResponse: {
          headers: new Headers(),
          ok: true,
          redirected: false,
          status: 200,
          statusText: 'OK',
          url: 'http://localhost',
          bodyUsed: false,
          body: {
            locked: false,
            cancel: jest.fn(),
            getReader: jest.fn(),
            pipeThrough: jest.fn(),
            pipeTo: jest.fn(),
            tee: jest.fn(),
          },
          blob: jest.fn(),
          arrayBuffer: jest.fn(),
          json: jest.fn(),
          text: jest.fn(),
          formData: jest.fn(),
          clone: jest.fn(),
          type: 'basic',
        },
      },
      mock,
    ]);

    return mock;
  }

  static findMockResponse(url: string, options: Readonly<RequestOptions>) {
    return Client.mockResponses.find(([response]) => {
      if (url !== response.url) {
        return false;
      }
      if ((options.method || 'GET') !== response.method) {
        return false;
      }
      return response.match.every(matcher => matcher(url, options));
    });
  }

  activeRequests: Record<string, typeof Request> = {};
  baseUrl = '';

  uniqueId() {
    return '123';
  }

  /**
   * In the real client, this clears in-flight responses. It's NOT
   * clearMockResponses. You probably don't want to call this from a test.
   */
  clear() {}

  wrapCallback<T extends any[]>(
    _id: string,
    func: FunctionCallback<T> | undefined,
    _cleanup: boolean = false
  ) {
    return (...args: T) => {
      if (hasProjectBeenRenamed(args)) {
        return;
      }
      respond(Client.mockAsync, func, ...args);
    };
  }

  requestPromise(
    path: string,
    {
      includeAllArgs,
      ...options
    }: {includeAllArgs?: boolean} & Readonly<RequestOptions> = {}
  ): any {
    return new Promise((resolve, reject) => {
      this.request(path, {
        ...options,
        success: (data, ...args) => {
          includeAllArgs ? resolve([data, ...args]) : resolve(data);
        },
        error: (error, ..._args) => {
          reject(error);
        },
      });
    });
  }

  static errors: Record<string, Error> = {};

  // XXX(ts): We type the return type for requestPromise and request as `any`. Typically these woul
  request(url: string, options: Readonly<RequestOptions> = {}): any {
    const [response, mock] = Client.findMockResponse(url, options) || [
      undefined,
      undefined,
    ];
    if (!response || !mock) {
      const methodAndUrl = `${options.method || 'GET'} ${url}`;
      // Endpoints need to be mocked
      const err = new Error(`No mocked response found for request: ${methodAndUrl}`);

      // Mutate stack to drop frames since test file so that we know where in the test
      // this needs to be mocked
      const lines = err.stack?.split('\n');
      const startIndex = lines?.findIndex(line => line.includes('.spec.'));
      err.stack = ['\n', lines?.[0], ...(lines?.slice(startIndex) ?? [])].join('\n');

      // Throwing an error here does not do what we want it to do....
      // Because we are mocking an API client, we generally catch errors to show
      // user-friendly error messages, this means in tests this error gets gobbled
      // up and developer frustration ensues.
      // We track the errors on a static member and warn afterEach test.
      Client.errors[methodAndUrl] = err;
    } else {
      // has mocked response

      // mock gets returned when we add a mock response, will represent calls to api.request
      mock(url, options);

      const body =
        typeof response.body === 'function' ? response.body(url, options) : response.body;

      if (![200, 202].includes(response.statusCode)) {
        response.callCount++;

        const errorResponse = Object.assign(
          {
            status: response.statusCode,
            responseText: JSON.stringify(body),
            responseJSON: body,
          },
          {
            overrideMimeType: () => {},
            abort: () => {},
            then: () => {},
            error: () => {},
          },
          new XMLHttpRequest()
        );

        this.handleRequestError(
          {
            id: '1234',
            path: url,
            requestOptions: options,
          },
          errorResponse as any,
          'error',
          'error'
        );
      } else {
        response.callCount++;
        respond(
          Client.mockAsync,
          options.success,
          body,
          {},
          {
            getResponseHeader: (key: string) => response.headers[key],
            statusCode: response.statusCode,
            status: response.statusCode,
          }
        );
      }
    }

    respond(Client.mockAsync, options.complete);
  }

  handleRequestError(
    {id, path, requestOptions}: HandleRequestErrorOptions,
    response: ResponseMeta,
    textStatus: string,
    errorThrown: string
  ) {
    const code = response?.responseJSON?.detail?.code;
    const isSudoRequired = code === SUDO_REQUIRED || code === SUPERUSER_REQUIRED;

    let didSuccessfullyRetry = false;

    if (isSudoRequired) {
      openSudo({
        isSuperuser: code === SUPERUSER_REQUIRED,
        sudo: code === SUDO_REQUIRED,
        retryRequest: async () => {
          try {
            const data = await this.requestPromise(path, requestOptions);
            requestOptions.success?.(data);
            didSuccessfullyRetry = true;
          } catch (err) {
            requestOptions.error?.(err);
          }
        },
        onClose: () =>
          // If modal was closed, then forward the original response
          !didSuccessfullyRetry && requestOptions.error?.(response),
      });
      return;
    }

    const errorCb = this.wrapCallback<[ResponseMeta, string, string]>(
      id,
      requestOptions.error
    );
    errorCb?.(response, textStatus, errorThrown);
  }
}

async function openSudo({onClose, ...args}: any = {}) {
  const mod = await import('sentry/components/modals/sudoModal');
  const {default: Modal} = mod;

  openModal(deps => <Modal {...deps} {...args} />, {onClose});
}

function openModal(renderer: (renderProps: any) => React.ReactNode, options?: any) {
  ModalStore.openModal(renderer, options ?? {});
}

export {Client};
