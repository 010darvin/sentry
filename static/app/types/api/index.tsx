import {paths} from 'sentry/openapi';

export type ExtractPathResponseData<
  TPath extends keyof paths,
  TOperation extends keyof paths[TPath]
> = paths[TPath][TOperation] extends {
  responses: {200: {content: {'application/json': any}}};
}
  ? paths[TPath][TOperation]['responses'][200]['content']['application/json']
  : never;
