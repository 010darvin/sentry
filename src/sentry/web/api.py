from django.http import HttpResponse
from django.views.decorators.cache import cache_control
from django.views.generic.base import View as BaseView
from rest_framework.request import Request
from rest_framework.response import Response

from sentry.utils import json
from sentry.web.client_config import get_client_config


class ClientConfigView(BaseView):
    def get(self, request: Request) -> Response:
        return HttpResponse(json.dumps(get_client_config(request)), content_type="application/json")


@cache_control(max_age=3600, public=True)
def robots_txt(request):
    return HttpResponse("User-agent: *\nDisallow: /\n", content_type="text/plain")
