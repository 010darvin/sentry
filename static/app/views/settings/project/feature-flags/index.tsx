import {Fragment, useEffect, useState} from 'react';
import styled from '@emotion/styled';
import isEqual from 'lodash/isEqual';

import {
  addErrorMessage,
  addLoadingMessage,
  addSuccessMessage,
} from 'sentry/actionCreators/indicator';
import {ModalRenderProps, openModal} from 'sentry/actionCreators/modal';
import Button from 'sentry/components/button';
import EmptyStateWarning from 'sentry/components/emptyStateWarning';
import FeatureBadge from 'sentry/components/featureBadge';
import CompactSelect from 'sentry/components/forms/compactSelect';
import ExternalLink from 'sentry/components/links/externalLink';
import SearchBar from 'sentry/components/searchBar';
import SentryDocumentTitle from 'sentry/components/sentryDocumentTitle';
import {IconAdd} from 'sentry/icons/iconAdd';
import {t, tct} from 'sentry/locale';
import ProjectStore from 'sentry/stores/projectsStore';
import space from 'sentry/styles/space';
import {Project} from 'sentry/types';
import {defined} from 'sentry/utils';
import handleXhrErrorResponse from 'sentry/utils/handleXhrErrorResponse';
import useApi from 'sentry/utils/useApi';
import useOrganization from 'sentry/utils/useOrganization';
import usePrevious from 'sentry/utils/usePrevious';
import SettingsPageHeader from 'sentry/views/settings/components/settingsPageHeader';
import TextBlock from 'sentry/views/settings/components/text/textBlock';
import PermissionAlert from 'sentry/views/settings/organization/permissionAlert';

import {FlagModal} from './modals/flagModal';
import {SegmentModal} from './modals/segmentModal';
import {Card} from './card';
import {Promo} from './promo';

type Props = ModalRenderProps & {
  project: Project;
};

export default function ProjectFeatureFlags({project}: Props) {
  const api = useApi();
  const organization = useOrganization();

  const currentFlags = project.featureFlags;
  const hasAccess = organization.access.includes('project:write');
  const disabled = !hasAccess;
  const previousFlags = usePrevious(currentFlags);

  const [flags, setFlags] = useState(currentFlags ?? {});
  const [query, setQuery] = useState('');
  const [status, setStatus] = useState(null);

  const showPromo = Object.keys(flags).length === 0;

  useEffect(() => {
    if (!isEqual(currentFlags, previousFlags)) {
      setFlags(currentFlags ?? {});
    }
  }, [currentFlags, previousFlags]);

  function handleAddFlag() {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
      />
    ));
  }

  function handleEditFlag(flagKey: string) {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  async function handleDeleteFlag(flagKey: string) {
    const newFeatureFlags = {...flags};
    delete newFeatureFlags[flagKey];

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted feature flag'));
    } catch (error) {
      const message = t('Unable to delete feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleActivateToggle(flagKey: string) {
    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].enabled = !newFeatureFlags[flagKey].enabled;

    addLoadingMessage();

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully updated the feature flag'));
    } catch (error) {
      const message = t('Unable to update the feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  function handleAddSegment(flagKey: string) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  function handleEditSegment(flagKey: string, index: number) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
        segmentIndex={index}
      />
    ));
  }

  async function handleDeleteSegment(flagKey: string, index: number) {
    const newEvaluations = [...flags[flagKey].evaluation];
    newEvaluations.splice(index, 1);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = newEvaluations;

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted segment'));
    } catch (error) {
      const message = t('Unable to delete segment');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleSortSegments(flagKey: string, segmentIds: string[]) {
    const sortedSegments = segmentIds
      .map(segmentId =>
        flags[flagKey].evaluation.find(segment => String(segment.id) === segmentId)
      )
      .filter(defined);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = sortedSegments;

    setFlags(newFeatureFlags);

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully sorted segments'));
    } catch (error) {
      setFlags(previousFlags ?? []);
      const message = t('Unable to sort segments');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  const filteredFlags = Object.keys(flags).filter(key => {
    return (
      key.toLowerCase().includes(query.toLowerCase()) &&
      (status === null || flags[key].enabled === status)
    );
  });

  return (
    <SentryDocumentTitle title={t('Feature Flags')}>
      <Fragment>
        <SettingsPageHeader
          title={
            <Fragment>
              {t('Feature Flags')} <FeatureBadge type="beta" />
            </Fragment>
          }
          action={
            !showPromo && (
              <Button
                title={
                  disabled ? t('You do not have permission to add flags') : undefined
                }
                priority="primary"
                size="sm"
                icon={<IconAdd size="xs" isCircled />}
                onClick={handleAddFlag}
                disabled={disabled}
              >
                {t('Add Flag')}
              </Button>
            )
          }
        />
        <TextBlock>
          {tct(
            'Feature flags allow you to configure your code into different flavors by dynamically toggling certain functionality on and off. Learn more about feature flags in our [link:documentation].',
            {
              link: <ExternalLink href="" />,
            }
          )}
        </TextBlock>

        <PermissionAlert access={['project:write']} />

        {showPromo ? (
          <Promo hasAccess={hasAccess} onGetStarted={handleAddFlag} />
        ) : (
          <Content>
            <Filters>
              <SearchBar
                size="md"
                onChange={setQuery}
                query={query}
                placeholder={t('Search by flag name')}
              />
              <CompactSelect
                defaultValue={status}
                options={[
                  {value: null, label: t('All')},
                  {value: true, label: t('Active')},
                  {value: false, label: t('Inactive')},
                ]}
                value={status}
                onChange={value => setStatus(value.value)}
                triggerProps={{prefix: t('Status'), style: {width: '100%'}}}
              />
            </Filters>
            {!filteredFlags.length ? (
              <EmptyStateWarning>
                <p>{t('Sorry, no feature flag match your filters')}</p>
              </EmptyStateWarning>
            ) : (
              <div>
                {filteredFlags.map(filteredFlag => (
                  <Card
                    key={filteredFlag}
                    flagKey={filteredFlag}
                    kind={flags[filteredFlag].kind}
                    enabled={flags[filteredFlag].enabled}
                    description={flags[filteredFlag].description}
                    segments={flags[filteredFlag].evaluation}
                    onDelete={() => handleDeleteFlag(filteredFlag)}
                    onEdit={() => handleEditFlag(filteredFlag)}
                    onActivateToggle={() => handleActivateToggle(filteredFlag)}
                    onAddSegment={() => handleAddSegment(filteredFlag)}
                    onDeleteSegment={id => handleDeleteSegment(filteredFlag, id)}
                    onEditSegment={id => handleEditSegment(filteredFlag, id)}
                    hasAccess={hasAccess}
                    onSortSegments={({reorderedItems}) =>
                      handleSortSegments(filteredFlag, reorderedItems)
                    }
                  />
                ))}
              </div>
            )}
          </Content>
        )}
      </Fragment>
    </SentryDocumentTitle>
  );
}

const Content = styled('div')`
  display: grid;
  gap: ${space(2)};
`;

const Filters = styled('div')`
  display: grid;
  gap: ${space(1)};
  grid-template-columns: 1fr max-content;
`;
