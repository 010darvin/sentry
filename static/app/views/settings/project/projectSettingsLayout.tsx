import {cloneElement, isValidElement, useMemo} from 'react';
import {RouteComponentProps} from 'react-router';

import * as AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import useRouteAnalyticsParams from 'sentry/utils/useRouteAnalyticsParams';
import withOrganization from 'sentry/utils/withOrganization';
import ProjectContext from 'sentry/views/projects/projectContext';
import SettingsLayout from 'sentry/views/settings/components/settingsLayout';
import ProjectSettingsNavigation from 'sentry/views/settings/project/projectSettingsNavigation';

type Props = {
  children: React.ReactNode;
  organization: Organization;
} & RouteComponentProps<{orgId: string; projectId: string}, {}>;

type InnerProps = Props & {project: Project};

function InnerProjectSettingsLayout({
  params,
  routes,
  project,
  organization,
  children,
  ...props
}: InnerProps) {
  // set analytics params for route based analytics
  useRouteAnalyticsParams(
    // use memoized value to avoid re-rendering
    useMemo(
      () => ({
        project_id: project.id,
        project_platform: project.platform,
      }),
      [project.id, project.platform]
    )
  );
  return (
    <AppStoreConnectContext.Provider project={project} organization={organization}>
      <SettingsLayout
        params={params}
        routes={routes}
        {...props}
        renderNavigation={() => <ProjectSettingsNavigation organization={organization} />}
      >
        {children && isValidElement(children)
          ? cloneElement<any>(children, {organization, project})
          : children}
      </SettingsLayout>
    </AppStoreConnectContext.Provider>
  );
}

function ProjectSettingsLayout({params, ...props}: Props) {
  const {orgId, projectId} = params;

  return (
    <ProjectContext orgId={orgId} projectId={projectId}>
      {({project}) => <InnerProjectSettingsLayout {...{params, project, ...props}} />}
    </ProjectContext>
  );
}

export default withOrganization(ProjectSettingsLayout);
