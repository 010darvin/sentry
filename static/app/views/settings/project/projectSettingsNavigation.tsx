import {useContext} from 'react';

import AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import withProject from 'sentry/utils/withProject';
import SettingsNavigation from 'sentry/views/settings/components/settingsNavigation';
import getConfiguration from 'sentry/views/settings/project/navigationConfiguration';

type Props = {
  organization: Organization;
  project?: Project;
};

const ProjectSettingsNavigation = ({organization, project}: Props) => {
  const appStoreConnectContext = useContext(AppStoreConnectContext);

  const debugFilesNeedsReview = appStoreConnectContext
    ? Object.keys(appStoreConnectContext).some(
        // @ts-expect-error TS(2532) FIXME: Object is possibly 'undefined'.
        key => appStoreConnectContext[key].credentials.status === 'invalid'
      )
    : false;

  return (
    <SettingsNavigation
      navigationObjects={getConfiguration({project, organization, debugFilesNeedsReview})}
      access={new Set(organization.access)}
      features={new Set(organization.features)}
      organization={organization}
      project={project}
    />
  );
};

export default withProject(ProjectSettingsNavigation);
