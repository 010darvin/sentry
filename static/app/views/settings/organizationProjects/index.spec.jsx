import {Organization} from 'fixtures/js-stubs/organization';
import {Project} from 'fixtures/js-stubs/project';
import {routerContext} from 'fixtures/js-stubs/routerContext';

import {render, screen, userEvent} from 'sentry-test/reactTestingLibrary';

import {Client} from 'sentry/api';
import OrganizationProjectsContainer from 'sentry/views/settings/organizationProjects';

describe('OrganizationProjects', function () {
  let org;
  let project;
  let projectsGetMock;
  let statsGetMock;
  let projectsPutMock;
  const routerContext = routerContext();

  beforeEach(function () {
    project = Project();
    org = Organization();

    projectsGetMock = Client.addMockResponse({
      url: '/organizations/org-slug/projects/',
      body: [project],
    });

    statsGetMock = Client.addMockResponse({
      url: '/organizations/org-slug/stats/',
      body: [[[], 1]],
    });

    projectsPutMock = Client.addMockResponse({
      method: 'PUT',
      url: '/projects/org-slug/project-slug/',
    });
  });

  afterEach(function () {
    Client.clearMockResponses();
  });

  it('should render the projects in the store', function () {
    const {container} = render(
      <OrganizationProjectsContainer params={{orgId: org.slug}} location={{query: {}}} />
    );

    expect(container).toSnapshot();

    expect(screen.getByText('project-slug')).toBeInTheDocument();

    expect(projectsGetMock).toHaveBeenCalledTimes(1);
    expect(statsGetMock).toHaveBeenCalledTimes(1);
    expect(projectsPutMock).toHaveBeenCalledTimes(0);

    userEvent.click(screen.getByRole('button', {name: 'Bookmark Project'}));
    expect(
      screen.getByRole('button', {name: 'Bookmark Project', pressed: true})
    ).toBeInTheDocument();

    expect(projectsPutMock).toHaveBeenCalledTimes(1);
  });

  it('should search organization projects', function () {
    const searchMock = MockApiClient.addMockResponse({
      url: `/organizations/${org.slug}/projects/`,
      body: [],
    });
    render(
      <OrganizationProjectsContainer location={{query: {}}} params={{orgId: org.slug}} />,
      {context: routerContext}
    );

    const searchBox = screen.getByRole('textbox');

    userEvent.type(searchBox, project.slug);

    expect(searchMock).toHaveBeenLastCalledWith(
      `/organizations/${org.slug}/projects/`,
      expect.objectContaining({
        method: 'GET',
        query: {
          query: project.slug,
        },
      })
    );

    userEvent.type(searchBox, '{enter}');
    expect(routerContext.context.router.push).toHaveBeenCalledTimes(1);
  });
});
