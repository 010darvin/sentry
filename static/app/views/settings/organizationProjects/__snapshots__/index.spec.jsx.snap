// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationProjects should render the projects in the store 1`] = `
<div>
  <div
    class="css-q8xgg6-SettingsPageHeader e1nr98yc2"
  >
    <div
      class="css-1mfa9vw-TitleAndActions e1nr98yc8"
    >
      <div
        class="css-u3au9h-TitleWrapper e1nr98yc7"
      >
        <div
          class="css-18xqiae-Title e1nr98yc6"
        >
          <h4
            class="css-jz1c1u-HeaderTitle e12cn1y20"
          >
            Projects
          </h4>
        </div>
      </div>
      <div
        class="css-1srjsw5-Action e1nr98yc3"
      >
        <a
          aria-disabled="false"
          aria-label="Create Project"
          class="e1b8u3ky2 css-1wsjsdp-Link-StyledButton-getColors-getButtonStyles e1ycxor00"
          role="button"
        >
          <span
            class="css-at526y-InteractionStateLayer-InteractionStateLayer-InteractionStateLayer e1fwp9rx0"
            role="presentation"
          />
          <span
            class="css-1z0n9wg-ButtonLabel e1b8u3ky1"
          >
            <span
              class="css-1ywt59b-Icon e1b8u3ky0"
            >
              <svg
                fill="currentColor"
                height="12px"
                viewBox="0 0 16 16"
                width="12px"
              >
                <path
                  d="M11.28,8.75H4.72a.75.75,0,1,1,0-1.5h6.56a.75.75,0,1,1,0,1.5Z"
                />
                <path
                  d="M8,12a.76.76,0,0,1-.75-.75V4.72a.75.75,0,0,1,1.5,0v6.56A.76.76,0,0,1,8,12Z"
                />
                <path
                  d="M8,16a8,8,0,1,1,8-8A8,8,0,0,1,8,16ZM8,1.53A6.47,6.47,0,1,0,14.47,8,6.47,6.47,0,0,0,8,1.53Z"
                />
              </svg>
            </span>
            Create Project
          </span>
        </a>
      </div>
    </div>
  </div>
  <div
    class="css-nqzxf5-SearchWrapper e1wqwm893"
  >
    <form
      class="css-1eu18ac-Form eup3waa0"
    >
      <input
        class="search css-1ws5oe1-Input-inputStyles ekvub0m0"
        placeholder="Search Projects"
        value=""
      />
    </form>
  </div>
  <div
    class="css-2fj46w-Panel edxrl2k0"
  >
    <div
      class="css-1v7t0a0-PanelHeader-getPadding eknsyxy0"
    >
      Projects
    </div>
    <div
      class="css-pemm1r-PanelBody-textStyles eszvord0"
    >
      <div
        class="css-1phdsb6-PanelItem-GridPanelItem e1wqwm892"
      >
        <div
          class="css-f3hcme-ProjectListItemWrapper e1wqwm891"
        >
          <div
            class="css-1jb4yvr-Wrapper e3cl2ic0"
          >
            <button
              aria-disabled="false"
              aria-label="Bookmark Project"
              aria-pressed="false"
              class="e1lcpz8z0 css-1jvq0r8-StyledButton-getColors-getButtonStyles-BookmarkStarButton e1b8u3ky2"
              role="button"
              type="button"
            >
              <span
                class="css-vul34l-InteractionStateLayer-InteractionStateLayer-InteractionStateLayer e1fwp9rx0"
                role="presentation"
              />
              <span
                class="css-1z0n9wg-ButtonLabel e1b8u3ky1"
              >
                <span
                  class="css-161yn40-Icon e1b8u3ky0"
                >
                  <svg
                    fill="#80708F"
                    height="16px"
                    viewBox="0 0 16 16"
                    width="16px"
                  >
                    <path
                      d="M12.47,15.63a.73.73,0,0,1-.35-.09L8,13.38,3.88,15.54a.75.75,0,0,1-.79,0,.76.76,0,0,1-.3-.74l.79-4.59L.24,6.91A.75.75,0,0,1,.66,5.63L5.27,5,7.33.79a.78.78,0,0,1,1.34,0L10.73,5l4.61.67a.75.75,0,0,1,.42,1.28l-3.34,3.25.79,4.59a.76.76,0,0,1-.3.74A.79.79,0,0,1,12.47,15.63ZM8,11.78a.85.85,0,0,1,.35.08l3.12,1.65L10.88,10a.78.78,0,0,1,.21-.67L13.62,6.9l-3.49-.51A.74.74,0,0,1,9.56,6L8,2.82,6.44,6a.74.74,0,0,1-.57.41L2.38,6.9,4.91,9.36a.78.78,0,0,1,.21.67l-.59,3.48,3.12-1.65A.85.85,0,0,1,8,11.78Z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <a
              class="css-1irtykh-Link-StyledLink e11cq06e0"
            >
              <div
                class="css-6orsd0-Wrapper e1vo4o574"
              >
                <div
                  class="e1vo4o573 css-j5xni3-Wrapper-Wrapper-StyledAvatar e8mgi9q0"
                  direction="right"
                  max="1"
                  size="18"
                >
                  <img
                    class="css-da8t6h-StyledPlatformIcon e8mgi9q2"
                    height="18"
                    src="[object Object]"
                    style="border-radius: 3px;"
                    width="18"
                  />
                </div>
                <div
                  class="css-ajl783-DisplayNameAndDescription e1vo4o572"
                >
                  <span
                    class="css-12dq2et-DisplayName e1vo4o571"
                    data-test-id="badge-display-name"
                  >
                    <span
                      class="css-1exbshj-BadgeDisplayName-overflowEllipsis-BadgeDisplayName evxkoyb0"
                    >
                      project-slug
                    </span>
                  </span>
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          class="css-1cwjgo8-ProjectStatsGraphWrapper e1wqwm890"
        >
          <div
            class="css-qr61vd-ChartContainer-getTooltipStyles e1d3yu8b0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
